orchestrator/relayer/src/fee_manager.rs

```diff
@@ -0,0 +1,196 @@
+    // - Mode is AlwaysRelay
+    // - Mode is either API or File and the batch has a profitable cost
+    // - Mode is either API or File and the batch has been waiting to be sent more than GRAVITY_BATCH_SENDING_SECS secs
+    pub async fn can_send_batch(
+        &mut self,
+        estimated_cost: &GasCost,
+        batch_fee: &Erc20Token,
+        contract_address: &EthAddress,
+    ) -> bool {
+        match self.mode {
+            RelayerMode::AlwaysRelay => {
+                return true;
+            }
+            RelayerMode::File => {
+                if self.should_send_at_non_profitable_cost(contract_address) {
+                    return true
+                }
+                let token_price = match self.get_token_price(&batch_fee.token_contract_address).await {
+                    Ok(token_price) => token_price,
+                    Err(_) => return false,
+                };
+
+                let estimated_fee = estimated_cost.get_total();
+                let batch_value = batch_fee.amount * token_price;
+
+                info!("estimate cost is {}, batch value is {}", estimated_fee, batch_value);
+                return batch_value >= estimated_fee
+            }
+            RelayerMode::Api => {
+                let body = json!({
+                    "batchFee": {
+                        "amount": batch_fee.amount.to_string(),
+                        "tokenContractAddress": batch_fee.token_contract_address
+                    },
+                    "estimatedCost": {
+                        "gas": estimated_cost.gas.to_string(),
+                        "gasPrice": estimated_cost.gas_price.to_string()
+                    }}
+                );
+
+                return match Client::new()
+                    .post(self.relayer_api_url.as_str())
+                    .json(&body)
+                    .send().await {
+                    Ok(resp) =>
+                        match resp.json().await {
+                        Ok(json) => {
+                            let api_response: ApiResponse = json;
+                            return if api_response.can_send {
+                                true
+                            } else {
+                                // code 5 means that it is not profitable but limit has not been
+                                // exceeded or no addresses are blacklisted
+                                // in that case we check if we should send at non profitable cost
+                                api_response.reason_type == 5 &&
+                                    self.should_send_at_non_profitable_cost(contract_address)
+                            }
+                        }
+                        Err(err) => {
+                            error!("error on relayer api: {}", err);
+                            false
+                        }
+                    }
+                    Err(err) => {
+                        error!("error on relayer api: {}", err);
```

Comment:
not sure how good the pretty printed err is going to be... so if this doesn't propagate the error / give the context, perhaps the logging could distinguish them (I assume one of them is a connection error and the other one is a json response parsing error?)

Reply:
okay changed